### PR TITLE
Add config to periodically run GC.

### PR DESCRIPTION
**Pull request by asdacap**
- May help with some user with very large VM RAM missing attestation.
- Can't find a way to specify runtime config at runtime to limit GC heap usage.

## Changes:
- Add config to periodically trigger GC.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [X] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

- Ran smokes. Slightly reduced ram usage, but it could just be noise.

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Nethermind.Core.Timers;
+
+public static class FunctionalTimer
```
Comment (DemuirGos) [inline src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:7]: why not use [`PeriodicTimer`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.periodictimer?view=net-6.0) ?

Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:7]: I like this interface better. Call a function, thats it. No need to store the timer somewhere, or remember to dispose it.

**Comment by asdacap** [comment]
Sync time. No setting vs gc every 60 second.
- Teku 6 hours 20 minute vs 11 hours 10 minute (EPYC 7601).
- Prysm: 6 hours 10 minute vs 5 hours 40 minute.
- Lodestar: 5 hours 40 minute vs 6 hours 10 minute.
- Lighthouse: 7 hours 20 minute vs 7 hours 50 minute.

**Comment by asdacap** [comment]
Highest total memory, managed memory, (just eyeballing) no setting vs gc every 60s vs 12GB mem pressure.
- Lighthouse 19.6 GB, 10.8 GB vs 18.5 GB, 9.9 GB vs (not complete yet)
- Lodestar 20.2 GB, 9.04 GB vs 19 GB, 8.21 GB vs (not complete yet)
- Prysm: 19.8 Gb, 9.72 vs 18.6 GB, 9.22 GB vs (not complete yet)
- Teku: 18.1 GB, 6.91 vs 16 GB, 8.6 GB vs (not complete yet)

**Comment by LukaszRozmej** [comment]
1. There are ways to limit memory in .NET, can you experiment with them first? https://learn.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#heap-limit (multiple in the link, not only one linked directly)
2. Before #4313 , memory cache was inducing GC all the time (which was bad too).

> max total memory, managed memory for the last 12 hour, no setting vs gc every 60s vs 12GB mem pressure (sync not complete).
> 
>     * Lighthouse 21.2, 11 vs 20.3, 10.66 vs 17.13, 10.77
> 
>     * Lodestar 21.9, 8.8 vs 20.5, 8.15 vs 23.26, 9.28
> 
>     * Prysm: 21.6, 8.4 vs 21.1, 5.59 vs 17.3, 9.4
> 
>     * Teku: 19.53, 7.08 vs 17.1, 9.2 vs 18.77, 6.53

Can you make the numbers clearer I don't get which is which.

File: src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:7]: `PeriodicTimer` can be a local variable, no need to store it like here:

https://github.com/NethermindEth/nethermind/blob/c977179ce2176d36248083ff73ba15794b5d0d54/src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs#L120-L141

Though we should add `using` there.

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /// <summary>
+    /// Simple util to periodically run something. Interval does not include the time taken by the action itself, so
+    /// not accurate. Blocks until cancellation token is cancelled, swallows `TaskCancelledException` (even by action)
+    /// by design.
+    /// </summary>
+    /// <param name="interval"></param>
+    /// <param name="cancellationToken"></param>
+    /// <param name="action"></param>
+    public static async Task RunEvery(TimeSpan interval, CancellationToken cancellationToken, Action<CancellationToken> action)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:17]: Maybe a static extension method on action?

Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:17]: Static extension method?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:17]: all extension methods are static

**Comment by asdacap** [comment]
Can't find ways to set the config at runtime.

File: src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Timers/FunctionalTimer.cs:7]: But you still need a variable, you still need a custom loop.

**Comment by LukaszRozmej** [comment]
> Can't find ways to set the config at runtime.

setting `DOTNET_GCHeapHardLimit` env variable doesn't wok?

**Comment by asdacap** [comment]
Setting it via `Environment.SetEnvironmentVariable` does not do anything. In constrast, is I set it in the launchSettings with value 1000, it crash immediately.

**Comment by LukaszRozmej** [comment]
> Setting it via `Environment.SetEnvironmentVariable` does not do anything. In constrast, is I set it in the launchSettings with value 1000, it crash immediately.

I meant setting it outside of the application.

**Pull request closed without merging**